desktop/notification/notificationtest/fdo.go

```diff
@@ -0,0 +1,189 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package notificationtest
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/godbus/dbus"
+
+	"github.com/snapcore/snapd/dbusutil"
+)
+
+const (
+	fdoBusName    = "org.freedesktop.Notifications"
+	fdoObjectPath = "/org/freedesktop/Notifications"
+	fdoInterface  = "org.freedesktop.Notifications"
+)
+
+type FdoNotification struct {
+	ID      uint32
+	AppName string
+	Icon    string
+	Summary string
+	Body    string
+	Actions []string
+	Hints   map[string]dbus.Variant
+	Expires int32
+}
+
+type FdoServer struct {
+	conn *dbus.Conn
+	err  *dbus.Error
+
+	mu            sync.Mutex
+	lastID        uint32
+	notifications map[uint32]*FdoNotification
+}
+
+func NewFdoServer() (*FdoServer, error) {
+	conn, err := dbusutil.SessionBusPrivate()
+	if err != nil {
+		return nil, err
+	}
+
+	server := new(FdoServer)
+	server.conn = conn
+	server.notifications = make(map[uint32]*FdoNotification)
+	conn.Export(fdoApi{server}, fdoObjectPath, fdoInterface)
+
+	reply, err := conn.RequestName(fdoBusName, dbus.NameFlagDoNotQueue)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	if reply != dbus.RequestNameReplyPrimaryOwner {
+		conn.Close()
+		return nil, fmt.Errorf("cannot obtain bus name %q", fdoBusName)
+	}
+	return server, nil
+}
+
+func (server *FdoServer) Stop() error {
+	if _, err := server.conn.ReleaseName(fdoBusName); err != nil {
+		return err
+	}
+	return server.conn.Close()
+}
+
+func (server *FdoServer) SetError(err *dbus.Error) {
```

Comment:
Could you please add a comment about which functions are expected to return the error. For example, just reading this context, I would kind of expected `Stop` above to also check for `server.err`. I think it can avoid some head-scratching.